### PR TITLE
[SPARK-20868][CORE] UnsafeShuffleWriter should verify the position after FileChannel.transferTo

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Long time ago we fixed a [bug](https://issues.apache.org/jira/browse/SPARK-3948) in shuffle writer about `FileChannel.transferTo`. We were not very confident about that fix, so we added a position check after the writing, try to discover the bug earlier.

 However this checking is missing in the new `UnsafeShuffleWriter`, this PR adds it.

https://issues.apache.org/jira/browse/SPARK-18105 maybe related to that `FileChannel.transferTo` bug, hopefully we can find out the root cause after adding this position check.

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @joshRosan

**Comment by JoshRosen** [comment]
I'm reviewing this now. For reference, #2824 was the earlier PR this references.

**Comment by JoshRosen** [comment]
LGTM.

**Comment by JoshRosen** [comment]
Hmm, looks like the test failures are legitimate.

File: core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java [hunk 1]
```diff
@@ -425,14 +425,8 @@ void forceSorterToSpill() throws IOException {
           long bytesToTransfer = partitionLengthInSpill;
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java]: Minor nit: we don't need this `bytesToTransfer` anymore. We can remove this mutable variable and just use `partitionLengthInSpill` in its place.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -376,6 +356,31 @@ private[spark] object Utils extends Logging {
...
   }
 
+  def copyFileStreamNIO(input: FileChannel, output: FileChannel, bytesToCopy: Long): Unit = {
+    val initialPos = output.position()
+    var count = 0L
+    // In case transferTo method transferred less data than we have required.
+    while (count < bytesToCopy) {
+      count += input.transferTo(count, bytesToCopy - count, output)
+    }
+
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Can we add an `assert(count == bytesToCopy)` here, just to be safe?

File: core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java [hunk 3]
```diff
@@ -425,14 +425,8 @@ void forceSorterToSpill() throws IOException {
...
           final long writeStartTime = System.nanoTime();
-          while (bytesToTransfer > 0) {
-            final long actualBytesTransferred = spillInputChannel.transferTo(
-              spillInputChannelPositions[i],
-              bytesToTransfer,
-              mergedFileOutputChannel);
-            spillInputChannelPositions[i] += actualBytesTransferred;
-            bytesToTransfer -= actualBytesTransferred;
-          }
+          Utils.copyFileStreamNIO(spillInputChannel, mergedFileOutputChannel, bytesToTransfer);
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java]: This is not equivalent to the old code. The `copyFileStreamNIO` method is assuming that you're starting to transfer from position 0 in the input channel, which is only true on the first iteration of the outer loop.

I think you need to add a fourth argument to `copyFileStreamNIO` to specify the starting position of the input channel.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by jiangxb1987** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks for the review, merging to master/2.2/2.1/2.0!

**Pull request closed without merging**